<a href="https://colab.research.google.com/github/jtanzi56/Prediction-of-Product-Sales/blob/main/Sales_Prediction_jtanzi.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

##Prediction of Product Sales
Jacob Tanzi



##Project Overview


**Project 1 - Part 2 (Core):**

We will continue to work off of the GitHub repository you created last week for your first portfolio project. The goal of your portfolio projects are to have items to show potential employers that demonstrate your data science skills and competencies.

The first project will be a sales prediction for food items sold at various stores. The goal of this is to help the retailer understand the properties of products and outlets that play crucial roles in increasing sales.

For Part 2, your task is to use Pandas to load in the data and start exploring and cleaning it. As data scientists, about 80% of our time is spent on data preparation - so this is a very important step!

1) How many rows and columns?bold text

2) What are the datatypes of each variable?

3) Are there duplicates? If so, drop any duplicates.

4) Identify missing values.

5) Decide on how to address the missing values and do it! (This requires your judgement, so explain your choice).

6) Confirm that there are no missing values after addressing them.

7) Find and fix any inconsistent categories of data (example: fix cat, Cat, and cats so that they are consistent).

8) For any numerical columns, obtain the summary statistics of each (min, max, mean).

##Load and Inspect Data


In [10]:
import pandas as pd
import numpy as np
import seaborn as sns

salesp = "/content/drive/MyDrive/CodingDojo/01-Fundamentals/Week02/Data/sales_predictions_2023.csv"

In [3]:
#loading in as panda dataframe
df_sp = pd.read_csv(salesp)

In [4]:
#inspecting with head and info
df_sp.head()

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0           FDA15         9.30          Low Fat         0.016047   
1           DRC01         5.92          Regular         0.019278   
2           FDN15        17.50          Low Fat         0.016760   
3           FDX07        19.20          Regular         0.000000   
4           NCD19         8.93          Low Fat         0.000000   

               Item_Type  Item_MRP Outlet_Identifier  \
0                  Dairy  249.8092            OUT049   
1            Soft Drinks   48.2692            OUT018   
2                   Meat  141.6180            OUT049   
3  Fruits and Vegetables  182.0950            OUT010   
4              Household   53.8614            OUT013   

   Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                       1999      Medium               Tier 1   
1                       2009      Medium               Tier 3   
2                       1999      Medium               Tier 1   
3                       1998         NaN               Tier 3   
4                       1987        High               Tier 3   

         Outlet_Type  Item_Outlet_Sales  
0  Supermarket Type1          3735.1380  
1  Supermarket Type2           443.4228  
2  Supermarket Type1          2097.2700  
3      Grocery Store           732.3800  
4  Supermarket Type1           994.7052

In [11]:
df2_sp = df_sp.copy()

In [12]:
df_sp.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8523 entries, 0 to 8522
Data columns (total 12 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Item_Identifier            8523 non-null   object 
 1   Item_Weight                7060 non-null   float64
 2   Item_Fat_Content           8523 non-null   object 
 3   Item_Visibility            8523 non-null   float64
 4   Item_Type                  8523 non-null   object 
 5   Item_MRP                   8523 non-null   float64
 6   Outlet_Identifier          8523 non-null   object 
 7   Outlet_Establishment_Year  8523 non-null   int64  
 8   Outlet_Size                6113 non-null   object 
 9   Outlet_Location_Type       8523 non-null   object 
 10  Outlet_Type                8523 non-null   object 
 11  Item_Outlet_Sales          8523 non-null   float64
dtypes: float64(4), int64(1), object(7)
memory usage: 799.2+ KB


##Clean Data


In [14]:
#how many rows/col
df_sp.shape

(8523, 12)

In [15]:
# checking data types
df_sp.dtypes

Item_Identifier               object
Item_Weight                  float64
Item_Fat_Content              object
Item_Visibility              float64
Item_Type                     object
Item_MRP                     float64
Outlet_Identifier             object
Outlet_Establishment_Year      int64
Outlet_Size                   object
Outlet_Location_Type          object
Outlet_Type                   object
Item_Outlet_Sales            float64
dtype: object



*   no duplicate columns 
*   dtypes look good so far



In [17]:
#checking for duplicate rows
print(f'There are {df_sp.duplicated().sum()} duplicate rows.')
#checking for duplicate data
df_sp.duplicated().sum()

There are 0 duplicate rows.


0



*   No duplicate rows
*   No duplicate data



In [18]:
#checking for missing values
df_sp.isna().sum()

Item_Identifier                 0
Item_Weight                  1463
Item_Fat_Content                0
Item_Visibility                 0
Item_Type                       0
Item_MRP                        0
Outlet_Identifier               0
Outlet_Establishment_Year       0
Outlet_Size                  2410
Outlet_Location_Type            0
Outlet_Type                     0
Item_Outlet_Sales               0
dtype: int64

In [20]:
# checking % missing values
df_sp.isna().sum() / len(df_sp) * 100

Item_Identifier               0.000000
Item_Weight                  17.165317
Item_Fat_Content              0.000000
Item_Visibility               0.000000
Item_Type                     0.000000
Item_MRP                      0.000000
Outlet_Identifier             0.000000
Outlet_Establishment_Year     0.000000
Outlet_Size                  28.276428
Outlet_Location_Type          0.000000
Outlet_Type                   0.000000
Item_Outlet_Sales             0.000000
dtype: float64



*   Item weight missing 1463 for 17%
*   Outlet Size missing 2410 for 28%






1.   Drop rows with missing values
2.   replace missing values with 'Unknown'
3.   impute missing values using most frequent value

checking how many rows with missing values



In [21]:
## Diplays only rows with nan values
df_sp[df_sp.isna().any(axis = 1)]

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
3              FDX07        19.20          Regular         0.000000   
7              FDP10          NaN          Low Fat         0.127470   
8              FDH17        16.20          Regular         0.016687   
9              FDU28        19.20          Regular         0.094450   
18             DRI11          NaN          Low Fat         0.034238   
...              ...          ...              ...              ...   
8504           NCN18          NaN          Low Fat         0.124111   
8508           FDW31        11.35          Regular         0.043246   
8509           FDG45         8.10          Low Fat         0.214306   
8514           FDA01        15.00          Regular         0.054489   
8519           FDS36         8.38          Regular         0.046982   

                  Item_Type  Item_MRP Outlet_Identifier  \
3     Fruits and Vegetables  182.0950            OUT010   
7               Snack Foods  107.7622            OUT027   
8              Frozen Foods   96.9726            OUT045   
9              Frozen Foods  187.8214            OUT017   
18              Hard Drinks  113.2834            OUT027   
...                     ...       ...               ...   
8504              Household  111.7544            OUT027   
8508  Fruits and Vegetables  199.4742            OUT045   
8509  Fruits and Vegetables  213.9902            OUT010   
8514                 Canned   57.5904            OUT045   
8519           Baking Goods  108.1570            OUT045   

      Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
3                          1998         NaN               Tier 3   
7                          1985      Medium               Tier 3   
8                          2002         NaN               Tier 2   
9                          2007         NaN               Tier 2   
18                         1985      Medium               Tier 3   
...                         ...         ...                  ...   
8504                       1985      Medium               Tier 3   
8508                       2002         NaN               Tier 2   
8509                       1998         NaN               Tier 3   
8514                       2002         NaN               Tier 2   
8519                       2002         NaN               Tier 2   

            Outlet_Type  Item_Outlet_Sales  
3         Grocery Store           732.3800  
7     Supermarket Type3          4022.7636  
8     Supermarket Type1          1076.5986  
9     Supermarket Type1          4710.5350  
18    Supermarket Type3          2303.6680  
...                 ...                ...  
8504  Supermarket Type3          4138.6128  
8508  Supermarket Type1          2587.9646  
8509      Grocery Store           424.7804  
8514  Supermarket Type1           468.7232  
8519  Supermarket Type1           549.2850  

[3873 rows x 12 columns]



*   Too many rows with missing values to remove
*   will look into how many unique entries for Item weight.
*   will look for correlation in Outlet Size and other column.




In [22]:
df_sp.nunique()

Item_Identifier              1559
Item_Weight                   415
Item_Fat_Content                5
Item_Visibility              7880
Item_Type                      16
Item_MRP                     5938
Outlet_Identifier              10
Outlet_Establishment_Year       9
Outlet_Size                     3
Outlet_Location_Type            3
Outlet_Type                     4
Item_Outlet_Sales            3493
dtype: int64

In [23]:
df_sp.nunique()/ len(df_sp) * 100

Item_Identifier              18.291681
Item_Weight                   4.869178
Item_Fat_Content              0.058665
Item_Visibility              92.455708
Item_Type                     0.187727
Item_MRP                     69.670304
Outlet_Identifier             0.117330
Outlet_Establishment_Year     0.105597
Outlet_Size                   0.035199
Outlet_Location_Type          0.035199
Outlet_Type                   0.046932
Item_Outlet_Sales            40.983222
dtype: float64

In [28]:
# slicing weight and type to see any consistency
df_sp.loc[:, ['Item_Weight', 'Item_Type']] 

Item_Weight              Item_Type
0           9.300                  Dairy
1           5.920            Soft Drinks
2          17.500                   Meat
3          19.200  Fruits and Vegetables
4           8.930              Household
...           ...                    ...
8518        6.865            Snack Foods
8519        8.380           Baking Goods
8520       10.600     Health and Hygiene
8521        7.210            Snack Foods
8522       14.800            Soft Drinks

[8523 rows x 2 columns]

In [29]:
#creating filter for null values
null_Item_Weight = df_sp['Item_Weight'].isna()

In [30]:
df_sp[null_Item_Weight]

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
7              FDP10          NaN          Low Fat         0.127470   
18             DRI11          NaN          Low Fat         0.034238   
21             FDW12          NaN          Regular         0.035400   
23             FDC37          NaN          Low Fat         0.057557   
29             FDC14          NaN          Regular         0.072222   
...              ...          ...              ...              ...   
8485           DRK37          NaN          Low Fat         0.043792   
8487           DRG13          NaN          Low Fat         0.037006   
8488           NCN14          NaN          Low Fat         0.091473   
8490           FDU44          NaN          Regular         0.102296   
8504           NCN18          NaN          Low Fat         0.124111   

                  Item_Type  Item_MRP Outlet_Identifier  \
7               Snack Foods  107.7622            OUT027   
18              Hard Drinks  113.2834            OUT027   
21             Baking Goods  144.5444            OUT027   
23             Baking Goods  107.6938            OUT019   
29                   Canned   43.6454            OUT019   
...                     ...       ...               ...   
8485            Soft Drinks  189.0530            OUT027   
8487            Soft Drinks  164.7526            OUT027   
8488                 Others  184.6608            OUT027   
8490  Fruits and Vegetables  162.3552            OUT019   
8504              Household  111.7544            OUT027   

      Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
7                          1985      Medium               Tier 3   
18                         1985      Medium               Tier 3   
21                         1985      Medium               Tier 3   
23                         1985       Small               Tier 1   
29                         1985       Small               Tier 1   
...                         ...         ...                  ...   
8485                       1985      Medium               Tier 3   
8487                       1985      Medium               Tier 3   
8488                       1985      Medium               Tier 3   
8490                       1985       Small               Tier 1   
8504                       1985      Medium               Tier 3   

            Outlet_Type  Item_Outlet_Sales  
7     Supermarket Type3          4022.7636  
18    Supermarket Type3          2303.6680  
21    Supermarket Type3          4064.0432  
23        Grocery Store           214.3876  
29        Grocery Store           125.8362  
...                 ...                ...  
8485  Supermarket Type3          6261.8490  
8487  Supermarket Type3          4111.3150  
8488  Supermarket Type3          2756.4120  
8490      Grocery Store           487.3656  
8504  Supermarket Type3          4138.6128  

[1463 rows x 12 columns]

In [36]:

df_sp.groupby('Item_Weight')['Item_Type'].sum().sort_values(ascending=True)

Item_Weight
5.425                  Baking GoodsBaking GoodsBaking Goods
7.090      Baking GoodsBaking GoodsBaking GoodsBaking Goods
7.445      Baking GoodsBaking GoodsBaking GoodsBaking Goods
5.635      Baking GoodsBaking GoodsBaking GoodsBaking Goods
4.920     Baking GoodsBaking GoodsBaking GoodsBaking Goo...
                                ...                        
16.000    Starchy FoodsStarchy FoodsFrozen FoodsDairyDai...
12.300    Starchy FoodsStarchy FoodsHouseholdSnack Foods...
7.060     Starchy FoodsStarchy FoodsStarchy FoodsStarchy...
7.600     Starchy FoodsStarchy FoodsStarchy FoodsStarchy...
7.435     Starchy FoodsStarchy FoodsStarchy FoodsStarchy...
Name: Item_Type, Length: 415, dtype: object



*   Item types with missing Item Weights have too many different values to draw con-
*         IE- all nul weights are canned foods, therefore N/A could apply 
*   Will replace with unknown



In [37]:
df_sp['Item_Weight'] = df_sp['Item_Weight'].fillna("UNKNOWN")
df_sp['Item_Weight'].value_counts(dropna=False)

UNKNOWN    1463
12.15        86
17.6         82
13.65        77
11.8         76
           ... 
7.275         2
7.685         1
9.42          1
6.52          1
5.4           1
Name: Item_Weight, Length: 416, dtype: int64

In [38]:
df_sp[null_Item_Weight]

Item_Identifier Item_Weight Item_Fat_Content  Item_Visibility  \
7              FDP10     UNKNOWN          Low Fat         0.127470   
18             DRI11     UNKNOWN          Low Fat         0.034238   
21             FDW12     UNKNOWN          Regular         0.035400   
23             FDC37     UNKNOWN          Low Fat         0.057557   
29             FDC14     UNKNOWN          Regular         0.072222   
...              ...         ...              ...              ...   
8485           DRK37     UNKNOWN          Low Fat         0.043792   
8487           DRG13     UNKNOWN          Low Fat         0.037006   
8488           NCN14     UNKNOWN          Low Fat         0.091473   
8490           FDU44     UNKNOWN          Regular         0.102296   
8504           NCN18     UNKNOWN          Low Fat         0.124111   

                  Item_Type  Item_MRP Outlet_Identifier  \
7               Snack Foods  107.7622            OUT027   
18              Hard Drinks  113.2834            OUT027   
21             Baking Goods  144.5444            OUT027   
23             Baking Goods  107.6938            OUT019   
29                   Canned   43.6454            OUT019   
...                     ...       ...               ...   
8485            Soft Drinks  189.0530            OUT027   
8487            Soft Drinks  164.7526            OUT027   
8488                 Others  184.6608            OUT027   
8490  Fruits and Vegetables  162.3552            OUT019   
8504              Household  111.7544            OUT027   

      Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
7                          1985      Medium               Tier 3   
18                         1985      Medium               Tier 3   
21                         1985      Medium               Tier 3   
23                         1985       Small               Tier 1   
29                         1985       Small               Tier 1   
...                         ...         ...                  ...   
8485                       1985      Medium               Tier 3   
8487                       1985      Medium               Tier 3   
8488                       1985      Medium               Tier 3   
8490                       1985       Small               Tier 1   
8504                       1985      Medium               Tier 3   

            Outlet_Type  Item_Outlet_Sales  
7     Supermarket Type3          4022.7636  
18    Supermarket Type3          2303.6680  
21    Supermarket Type3          4064.0432  
23        Grocery Store           214.3876  
29        Grocery Store           125.8362  
...                 ...                ...  
8485  Supermarket Type3          6261.8490  
8487  Supermarket Type3          4111.3150  
8488  Supermarket Type3          2756.4120  
8490      Grocery Store           487.3656  
8504  Supermarket Type3          4138.6128  

[1463 rows x 12 columns]

In [43]:
null_sums = df_sp.isna().sum()

null_sums

Item_Identifier                 0
Item_Weight                     0
Item_Fat_Content                0
Item_Visibility                 0
Item_Type                       0
Item_MRP                        0
Outlet_Identifier               0
Outlet_Establishment_Year       0
Outlet_Size                  2410
Outlet_Location_Type            0
Outlet_Type                     0
Item_Outlet_Sales               0
dtype: int64

In [44]:
df_sp.loc[:, ['Outlet_Size', 'Outlet_Type']] 

Outlet_Size        Outlet_Type
0         Medium  Supermarket Type1
1         Medium  Supermarket Type2
2         Medium  Supermarket Type1
3            NaN      Grocery Store
4           High  Supermarket Type1
...          ...                ...
8518        High  Supermarket Type1
8519         NaN  Supermarket Type1
8520       Small  Supermarket Type1
8521      Medium  Supermarket Type2
8522       Small  Supermarket Type1

[8523 rows x 2 columns]

In [46]:
null_Outlet_Size = df_sp['Outlet_Size'].isna()
null_Outlet_Size

0       False
1       False
2       False
3        True
4       False
        ...  
8518    False
8519     True
8520    False
8521    False
8522    False
Name: Outlet_Size, Length: 8523, dtype: bool

In [49]:
df_sp.loc[:, ['Outlet_Size', 'Outlet_Establishment_Year']] 

Outlet_Size  Outlet_Establishment_Year
0         Medium                       1999
1         Medium                       2009
2         Medium                       1999
3            NaN                       1998
4           High                       1987
...          ...                        ...
8518        High                       1987
8519         NaN                       2002
8520       Small                       2004
8521      Medium                       2009
8522       Small                       1997

[8523 rows x 2 columns]

In [47]:
df_sp[null_Outlet_Size]

Item_Identifier Item_Weight Item_Fat_Content  Item_Visibility  \
3              FDX07        19.2          Regular         0.000000   
8              FDH17        16.2          Regular         0.016687   
9              FDU28        19.2          Regular         0.094450   
25             NCD06        13.0          Low Fat         0.099887   
28             FDE51       5.925          Regular         0.161467   
...              ...         ...              ...              ...   
8502           NCH43        8.42          Low Fat         0.070712   
8508           FDW31       11.35          Regular         0.043246   
8509           FDG45         8.1          Low Fat         0.214306   
8514           FDA01        15.0          Regular         0.054489   
8519           FDS36        8.38          Regular         0.046982   

                  Item_Type  Item_MRP Outlet_Identifier  \
3     Fruits and Vegetables  182.0950            OUT010   
8              Frozen Foods   96.9726            OUT045   
9              Frozen Foods  187.8214            OUT017   
25                Household   45.9060            OUT017   
28                    Dairy   45.5086            OUT010   
...                     ...       ...               ...   
8502              Household  216.4192            OUT045   
8508  Fruits and Vegetables  199.4742            OUT045   
8509  Fruits and Vegetables  213.9902            OUT010   
8514                 Canned   57.5904            OUT045   
8519           Baking Goods  108.1570            OUT045   

      Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
3                          1998         NaN               Tier 3   
8                          2002         NaN               Tier 2   
9                          2007         NaN               Tier 2   
25                         2007         NaN               Tier 2   
28                         1998         NaN               Tier 3   
...                         ...         ...                  ...   
8502                       2002         NaN               Tier 2   
8508                       2002         NaN               Tier 2   
8509                       1998         NaN               Tier 3   
8514                       2002         NaN               Tier 2   
8519                       2002         NaN               Tier 2   

            Outlet_Type  Item_Outlet_Sales  
3         Grocery Store           732.3800  
8     Supermarket Type1          1076.5986  
9     Supermarket Type1          4710.5350  
25    Supermarket Type1           838.9080  
28        Grocery Store           178.4344  
...                 ...                ...  
8502  Supermarket Type1          3020.0688  
8508  Supermarket Type1          2587.9646  
8509      Grocery Store           424.7804  
8514  Supermarket Type1           468.7232  
8519  Supermarket Type1           549.2850  

[2410 rows x 12 columns]

In [48]:
df_sp['Outlet_Size'].value_counts()

Medium    2793
Small     2388
High       932
Name: Outlet_Size, dtype: int64



*   No connection found between missing outlet size and Outlet_Location_Type or	Outlet_Type
*   Outlet size small and medium too close in value to make assumptions 
*   Will replace NaN with 'MISSING'



In [50]:
df_sp['Outlet_Size'] = df_sp['Outlet_Size'].fillna('MISSING')
df_sp['Outlet_Size'].value_counts(dropna=False)

Medium     2793
MISSING    2410
Small      2388
High        932
Name: Outlet_Size, dtype: int64

**Confirming no missing values after addressing them**

In [51]:
print(df_sp.isna().sum())

Item_Identifier              0
Item_Weight                  0
Item_Fat_Content             0
Item_Visibility              0
Item_Type                    0
Item_MRP                     0
Outlet_Identifier            0
Outlet_Establishment_Year    0
Outlet_Size                  0
Outlet_Location_Type         0
Outlet_Type                  0
Item_Outlet_Sales            0
dtype: int64


**Find and fix any inconsistent categories of data**

In [52]:
data_types = df_sp.dtypes
str_cols = data_types[data_types=='object'].index
str_cols

Index(['Item_Identifier', 'Item_Weight', 'Item_Fat_Content', 'Item_Type',
       'Outlet_Identifier', 'Outlet_Size', 'Outlet_Location_Type',
       'Outlet_Type'],
      dtype='object')

In [53]:
for col in str_cols:
    print(f'- {col}:')
    print(df_sp[col].value_counts(dropna=False))
    print("\n\n")

- Item_Identifier:
FDW13    10
FDG33    10
NCY18     9
FDD38     9
DRE49     9
         ..
FDY43     1
FDQ60     1
FDO33     1
DRF48     1
FDC23     1
Name: Item_Identifier, Length: 1559, dtype: int64



- Item_Weight:
UNKNOWN    1463
12.15        86
17.6         82
13.65        77
11.8         76
           ... 
7.275         2
7.685         1
9.42          1
6.52          1
5.4           1
Name: Item_Weight, Length: 416, dtype: int64



- Item_Fat_Content:
Low Fat    5089
Regular    2889
LF          316
reg         117
low fat     112
Name: Item_Fat_Content, dtype: int64



- Item_Type:
Fruits and Vegetables    1232
Snack Foods              1200
Household                 910
Frozen Foods              856
Dairy                     682
Canned                    649
Baking Goods              648
Health and Hygiene        520
Soft Drinks               445
Meat                      425
Breads                    251
Hard Drinks               214
Others                    169
Starchy Foods 

**Item_Fat_Content has inconsistencies**


1.   LF and low fat should be  Low Fat 
2.   reg should be Regular




In [54]:
Item_Fat_dict = {'low fat':'Low Fat',
                   'LF':'Low Fat',
                   'reg':'Regular'}

df_sp['Item_Fat_Content'] = df_sp['Item_Fat_Content'].replace(Item_Fat_dict)

In [55]:
df_sp['Item_Fat_Content'].value_counts()

Low Fat    5517
Regular    3006
Name: Item_Fat_Content, dtype: int64

#Summary Statistics for Numeric
- numerical columns, will obtain the summary statistics of each (min, max, mean).

In [56]:
df_sp.describe().round(2)

Item_Visibility  Item_MRP  Outlet_Establishment_Year  Item_Outlet_Sales
count          8523.00   8523.00                    8523.00            8523.00
mean              0.07    140.99                    1997.83            2181.29
std               0.05     62.28                       8.37            1706.50
min               0.00     31.29                    1985.00              33.29
25%               0.03     93.83                    1987.00             834.25
50%               0.05    143.01                    1999.00            1794.33
75%               0.09    185.64                    2004.00            3101.30
max               0.33    266.89                    2009.00           13086.96

In [58]:
#mean min and max for numeric column 
stats =  df_sp.describe()
stats.loc[['mean','min','max']]

Item_Visibility    Item_MRP  Outlet_Establishment_Year  \
mean         0.066132  140.992782                1997.831867   
min          0.000000   31.290000                1985.000000   
max          0.328391  266.888400                2009.000000   

      Item_Outlet_Sales  
mean        2181.288914  
min           33.290000  
max        13086.964800

##Exploratory Data Analysis


##Explanatory Data Analysis